OpenRA.Mods.Common/Traits/Production/ProductionByDelivery.cs

```diff
@@ -0,0 +1,223 @@
+			});
+
+			return true;
+		}
+
+		void TryToLand(Actor deliveringActor, IEnumerable<ActorInfo> actorsToProduce)
+		{
+			// Check if there is a valid drop-off point before beginning to descend
+			var exit = GetAvailableExit(self, actorsToProduce.First());
+
+			// Abort the landing and refund the player
+			if (exit == null || self.IsDead || !self.IsInWorld)
+			{
+				var value = actorsToProduce.Sum(actor => actor.Traits.Get<ValuedInfo>().Cost);
+				self.Owner.PlayerActor.Trait<PlayerResources>().GiveCash(value);
+
+				if (isPlane)
+					deliveringActor.QueueActivity(new Fly(deliveringActor, Target.FromCell(self.World, endPos)));
+				else
+					deliveringActor.QueueActivity(new HeliFly(deliveringActor, Target.FromCell(self.World, endPos)));
+
+				deliveringActor.QueueActivity(new RemoveSelf());
+
+				return;
+			}
+
+			if (isPlane)
+				deliveringActor.QueueActivity(new Land(deliveringActor, Target.FromActor(self)));
+			else
+			{
+				deliveringActor.QueueActivity(new HeliFly(deliveringActor, Target.FromActor(self)));
+				deliveringActor.QueueActivity(new HeliLand(deliveringActor, false));
+			}
+
+			deliveringActor.QueueActivity(new CallFunc(() => MakeDelivery(actorsToProduce.ToList(), deliveringActor, deliveringActor.World)));
+		}
+
+		void MakeDelivery(ICollection<ActorInfo> actorInfos, Actor deliveringActor, World world)
+		{
+			if (!actorInfos.Any())
+			{
+				if (isPlane)
+					deliveringActor.QueueActivity(new Fly(deliveringActor, Target.FromCell(world, endPos)));
+				else
+					deliveringActor.QueueActivity(new HeliFly(deliveringActor, Target.FromCell(world, endPos)));
+
+				deliveringActor.QueueActivity(new RemoveSelf());
+			}
+			else
+			{
+				var actorInfo = actorInfos.First();
+
+				var chosenExit = GetAvailableExit(self, actorInfo);
+				if (chosenExit == null)
+				{
+					WaitAndRetry(deliveringActor, world, actorInfos);
+					return;
+				}
+
+				var exitLocation = self.Location + chosenExit.ExitCell;
+				var targetLocation = RallyPoint.Value != null ? RallyPoint.Value.Location : exitLocation;
+
+				var newActor = DoProduction(self, actorInfo, chosenExit, factionVariant);
+
+				actorInfos.Remove(actorInfo);
+
+				world.AddFrameEndTask(w =>
+				{
+					MoveIntoWorld(w, newActor, chosenExit, exitLocation, targetLocation);
+					IssueNotifications(self, newActor, exitLocation);
+
+					deliveringActor.QueueActivity(new CallFunc(() => MakeDelivery(actorInfos, deliveringActor, world)));
+					Sound.PlayNotification(self.World.Map.Rules, self.Owner, "Speech", info.ReadyAudio, self.Owner.Country.Race);
+				});
+			}
+		}
+
+		void WaitAndRetry(Actor actor, World world, ICollection<ActorInfo> actorInfos)
+		{
+			actor.QueueActivity(new Wait(10));
+			actor.QueueActivity(new CallFunc(() => MakeDelivery(actorInfos, actor, world)));
+		}
+
+		static void IssueNotifications(Actor self, Actor newActor, CPos exitLocation)
```

Comment:
Could you move this in to Production (and then maybe rename it to something like `AfterProduction`) to reduce duplication?